### PR TITLE
lots of boilerplate

### PR DESCRIPTION
**Pull request by gress2**

### CODE REVIEW

File: ext/bit/bit_details.hpp
Comment (bkille) [inline ext/bit/bit_details.hpp]:
```suggestion
template <class InputIter>
```

Comment (bkille) [inline ext/bit/bit_details.hpp]:
```suggestion
template <class InputIt>
```

File: ext/bit/bit_details.hpp
Comment (bkille) [inline ext/bit/bit_details.hpp]:
```suggestion
range_type get_range_type(bit_iterator<InputIt> first, bit_iterator<WrappedIter> last) {
```

Following convention used in Vincent's existing code. I also think that using "right" and "left" should be avoided if possible.

Comment (bkille) [inline ext/bit/bit_details.hpp]: I think the terminology of "RIGHT" and "LEFT" can be confusing. For example, instead of "RIGHT_ALIGNED" and "LEFT_ALIGNED", we could have  "LSB_ALIGNED" and "LSB_UNALIGNED".

File: ext/bit/bit_details.hpp
Comment (bkille) [inline ext/bit/bit_details.hpp]: Should we create a separate file for debugging utilities like this?

File: include/accumulate.hpp
Comment (bkille) [inline include/accumulate.hpp:10]: Should be another // ==== // line followed by 3 newlines and _then_ the preamble header.

File: include/accumulate.hpp
Comment (bkille) [inline include/accumulate.hpp:18]:
```suggestion
Miscellaneous
```

File: include/accumulate.hpp
Comment (bkille) [inline include/accumulate.hpp:20]:
```suggestion
```

File: include/accumulate.hpp
Comment (bkille) [inline include/accumulate.hpp:38]:
```suggestion
```

File: include/count.hpp
Comment (bkille) [inline include/count.hpp]: I'm not sure if this is just GitHub's display, but all indents should be 4 spaces wide.

File: include/fill.hpp
Comment (bkille) [inline include/fill.hpp]: Same issue with "WrappedIter"

File: include/fill.hpp
Comment (bkille) [inline include/fill.hpp]: I believe you already do `last-1` in your implementation of `get_range_type(lhs, rhs)`

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: I think this function should be renamed. The use of "range" implies that we can use it over any range of valid iterators. I think it would be better to name it `set_within_word` or something similar.

Comment (bkille) [inline include/bit_algorithm_details.hpp]:
```suggestion
void set_in_range(BitIter first, BitIter last, 
```

File: include/fill_n.hpp
Comment (bkille) [inline include/fill_n.hpp]: Are we allowed to add default arguments to the `std` signature?

File: include/fill_n.hpp
Comment (bkille) [inline include/fill_n.hpp]: We cannot use  this with an Input/Output iterator

File: include/mismatch.hpp
Comment (bkille) [inline include/mismatch.hpp]: Again I think we should use `first1, last1, first2, last2` to follow the std.

File: include/mismatch.hpp
Comment (bkille) [inline include/mismatch.hpp]: I'm assuming this was for debugging. I'll skip reviewing this till the next PR

File: include/sample.hpp
Comment (bkille) [inline include/sample.hpp]: I think it's important to consider the iterator types here. For starters, we may not be able to use `distance`.

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: We should decide on a consistent naming of `WordType`. I agree that it makes sense, but `T` is less verbose and also more popular in the STL.

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: Why use `std::distance` instead of `last.position()-first.position()`?

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: Is this assignment necessary? Why can't we just use `.position()`

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: I think the logic here should be condensed. It seems odd to left shift by some `n-1` just to left shift it again right after by `1`

File: include/fill_n.hpp
Comment (bkille) [inline include/fill_n.hpp:26]: Are we allowed to increment `first` if it is an `OutputIt`?

File: ext/bit/bit_details.hpp
Comment (bkille) [inline ext/bit/bit_details.hpp]: Should we change the name to imply that the queried range must be within a word?

**Review by bkille — CHANGES_REQUESTED** [review]
I don't have any conceptual issues with the PR, but a I do have a few formatting/naming/convention concerns that I feel should be addressed before a merge. Also, I think that any files that are still in the debugging phase should be left as a stub (`mismatch`) still had `exit(1)`. 

### TODO before pull
- WrappedIt should be replaced with the corresponding iterator type.
- Format needs to strictly follow the rules in the README (newlines, tabs etc)
- Debugging utlities should be moved to their own file inside `include/`. I saw some printing stuff in bit_details.hpp.
- Avoid using "right" and "left" (This is up for debate. Right/Left shift terminology is unavoidable, but I think I would prefer "LSB aligned" or "MSB aligned" or something like that. 
- Internal functions like `set_in_range` and `get_range_type` I think should be prefixed by `_`.


### General comments
- Should we use `WordType` or `T` when dealing with the words?
- I believe we have some overlap in developed utility functions (my fault for not PRing). I will submit a PR so we can resolve this.
- We should decide sooner rather than later who will be implementing what for the near future at least.
- We should have a base set of tests that a file should have to pass to be committed, i.e. it should word for all valid iterator types, words, and alignments. Catch2 can handle the first 2 for us, so we only need to create a product test template and inside of it handle the alignments. I don't think that single word tests are sufficient for a PR (IMO)
- This is my first PR review. If I am being to harsh please let me know ):

File: ext/bit/bit_details.hpp
Comment (gress2) [reply ext/bit/bit_details.hpp]: Agreed.

File: ext/bit/bit_details.hpp
Comment (gress2) [reply ext/bit/bit_details.hpp]: Missed this. Actually, going to move this stuff out of ext/ as I don't really want to mess with the bit library itself.

File: ext/bit/bit_details.hpp
Comment (gress2) [reply ext/bit/bit_details.hpp]: Agreed.

File: ext/bit/bit_details.hpp
Comment (gress2) [reply ext/bit/bit_details.hpp]: Likely, yes.

File: include/accumulate.hpp
Comment (gress2) [reply include/accumulate.hpp:10]: Will fix this.

File: ext/bit/bit_details.hpp
Comment (gress2) [reply ext/bit/bit_details.hpp]: Removing this range_type altogether. I think it'll end up doing more harm than good.

File: include/accumulate.hpp
Comment (gress2) [reply include/accumulate.hpp:20]: leaving this for header contributors to fix.

File: include/accumulate.hpp
Comment (gress2) [reply include/accumulate.hpp:38]: leaving this for header contributors to fix.

**Comment by gress2** [comment]
I've addressed the issues here and wen't ahead and pulled your latest stuff into the branch and (attempted to) resolve merge conflicts.

File: include/accumulate.hpp
Comment (bkille) [inline include/accumulate.hpp:12]: I believe there should be 3 newlines here.

File: include/binary_search.hpp
Comment (bkille) [inline include/binary_search.hpp:18]: no newline before bit I think

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp:22]: Preamble should have separator line after namespace followed by 3 newlines.

File: ext/bit/bit_details.hpp
Comment (bkille) [inline ext/bit/bit_details.hpp]: Why are these 4 lines deleted?

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp:2]: I think these are all supposed to be aligned

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: Just a note: I think this function should be removed, or, preferably modified to account for the case when we are writing to more than 2 bits.

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: I think this is a useful function, but I think it should follow the style of _bitblend in the sense that it should take a word, two numbers (`start` and `len`) and also a `bv`

**Review by bkille — COMMENTED** [review]
I think I still have some questions/thoughts about all of this, but I'll leave it open for discussion this time instead of requesting changes. I'll let @vreverdy take a look and approve or request more changes. 

- `// ------ //` followed by 3 newlines before the `#endif` and namespace closing. I'm not sure why these are removed in the commit, perhaps I misunderstood the documentation.
- Some files look like they have too many newlines within blocks (preamble and closing section) and there are not enough newlines between blocks.
- I think we should sit down and discuss the testing utilities in more detail. Do we really need to deal with strings at all? 
- I also think we should discuss the iterator/algorithm utilities. I think some of them may be useful, but should be slightly modified. For some, we should decide which implementation is better, since some already exist.
- `_read_word_raw` and `_write_word_raw` should be recreated to work with multiple words and not at most 2. I think `get_word` is the same as read word, so we should decide which framework is better. Yours might end up being much more concise. same with write word.

File: include/bit_algorithm_details.hpp
Comment (gress2) [reply include/bit_algorithm_details.hpp]: Maybe, but this is a modifying operation while `_bitblend` is not.

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: How is this different than `_shld`? It seems like the only difference is that this assumes that `_shift_and_merge` guarantees that `w1_position` is within the word and `_shld` does not. If this is the case, should we create an overload of `_shld` or something? It seems odd to have two functions with nearly identical functionality.

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: Maybe I am misremembering, but I thought at the last meeting we discussed that `distance` might not be able to be used on `InputIt`. Not only this, but this function still only handles the case where we are spanning at most two words. 

I think it's worth discussing at our next meeting. While we can have general logic for reading any `T1` bits from any `T2` bit iterator, it may prove faster if we know `T1` and `T2` are the same, then we know that we will need to read from at most 2 words.  If that is the case, though, we should decide on some naming and perhaps overload `_read_word_raw`

File: include/bit_algorithm_details.hpp
Comment (bkille) [inline include/bit_algorithm_details.hpp]: This is most likely an extremely minor thing, but is it worth thinking about avoiding the `if/else` condition and replacing it with a single statement? I don't know much about branching, but I figure if we can avoid if/else why not.

File: include/fill_n.hpp
Comment (bkille) [reply include/fill_n.hpp:26]: Would you mind explaining why this is possible? Otherwise we can leave the discussion open just until we meet tomorrow, if you don't mind, so that  I can ask about it then.

**Review by bkille — APPROVED** [review]

**Pull request merged**
